ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -31,6 +60,299 @@ let tdb_tool = !Xapi_globs.tdb_tool
+        parse_string ~consume:All kvp_map x
+    end in
+    let ldap fmt = fmt |> Printf.ksprintf @@ Printf.sprintf "ldap %s" in
+    let* kvps = P.parse_kvp_map stdout <!> ldap "parsing failed '%s'" in
+    let get_string k =
+      match Map.find_opt k kvps with
+      | None ->
+          Error (ldap "missing key '%s'" k)
+      | Some x ->
+          Ok x
+    in
+    let get_int of_string k =
+      let* str = get_string k in
+      try Ok (of_string str)
+      with _ -> Error (ldap "invalid value for key '%s'" k)
+    in
+    let* upn = get_string "userPrincipalName" in
+    let* user_account_control = get_int Int32.of_string "userAccountControl" in
+    let* account_expires = get_int Int64.of_string "accountExpires" in
+    let account_expired =
+      (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-accountexpires *)
+      let windows_nt_time_to_unix_time x =
+        Int64.sub (Int64.div x 10000000L) 11644473600L
+      in
+      match account_expires with
+      | x when x = 0L || x = Int64.max_int ->
+          false
+      | i ->
+          let expire_unix_time =
+            windows_nt_time_to_unix_time i |> Int64.to_float
+          in
+          expire_unix_time < Unix.time ()
+    in
+    let open Int32 in
+    (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-useraccountcontrol#remarks
+     * for bit flag docs *)
+    let disabled_bit = of_string "0x2" in
+    let lockout_bit = of_string "0x10" in
+    let passw_expire_bit = of_string "0x800000" in
+    Ok
+      {
+        upn
+      ; account_expired
+      ; account_disabled= logand user_account_control disabled_bit <> 0l
+      ; account_locked= logand user_account_control lockout_bit <> 0l
+      ; password_expired= logand user_account_control passw_expire_bit <> 0l
+      }
+
+
+  let net_ads (sid : string) : (string, exn) result =
+    try
+      let args = ["ads"; "sid"; "-d"; debug_level; "--machine-pass"; sid] in
+      let stdout =
+        Helpers.call_script ~log_output:On_failure !Xapi_globs.net_cmd args
+      in
+      Ok stdout
+    with _ -> Error (generic_ex "net ads query failed")
+
+  let query_user sid =
+    let* stdout =
+      try
+        let args = ["ads"; "sid"; "-d"; debug_level; "--machine-pass"; sid] in
+        let stdout =
+          Helpers.call_script ~log_output:On_failure !Xapi_globs.net_cmd args
+        in
+        Ok stdout
+      with _ -> Error (generic_ex "ldap query failed")
+    in
+    parse_user stdout <!> generic_ex "%s"
+end
+
+module Wbinfo = struct
+
+  let exception_of_stderr =
+    let open Auth_signature in
+    let regex = Re.Perl.(compile (re {|.*(WBC_ERR_[A-Z_]*).*|})) in
+    let get_regex_match x =
+      Option.bind (Re.exec_opt regex x) (fun g ->
+          match Re.Group.all g with [|_; code|] -> Some code | _ -> None)
+    in
+    fun stderr ->
+      get_regex_match stderr
+      |> Option.map (fun code ->
+             match code with
+             | "WBC_ERR_AUTH_ERROR" ->
```

Comment:
I suppose this error list comes from https://github.com/samba-team/samba/blob/master/nsswitch/libwbclient/wbclient.h#L36

a comment to link it will be good.